### PR TITLE
`set_value_grad` propagate gradients to `Input` and `TensorValue` 

### PR DESCRIPTION
**Pull request by hbwx24**
### PR types

Function optimization
### PR changes

OPs
### Describe

set_value 梯度截断。

### CODE REVIEW

File: paddle/fluid/operators/set_value_op.h [hunk 1]
```diff
@@ -22,15 +22,35 @@
...
 #include "paddle/fluid/operators/utils.h"
 #include "paddle/fluid/platform/enforce.h"
 
 namespace paddle {
 namespace operators {
 
 using Tensor = framework::Tensor;
+using DDim = framework::DDim;
+
+inline void get_offsets(const DDim& big_dim, const DDim& small_dim,
```
Comment (chenwhql) [inline paddle/fluid/operators/set_value_op.h]: naming style, get_offsets -> GetOffsets

File: paddle/fluid/operators/set_value_op.h [hunk 2]
```diff
@@ -292,5 +312,255 @@ class SetValueKernel : public framework::OpKernel<T> {
...
+        break;
+      }
+    }
+
+    auto& dev_ctx = context.template device_context<DeviceContext>();
+    auto& place =
+        *context.template device_context<DeviceContext>().eigen_device();
+    math::SetConstant<DeviceContext, T> set_zero;
+
+    if ((grad_value) && (!grad_value->IsInitialized())) {
```
Comment (chenwhql) [inline paddle/fluid/operators/set_value_op.h]: why need this branch? grad_value can be initialized?

File: paddle/fluid/operators/set_value_op.h [hunk 3]
```diff
@@ -292,5 +312,255 @@ class SetValueKernel : public framework::OpKernel<T> {
...
+      framework::Tensor tmp(grad_input->type());
+      tmp.mutable_data<T>(out_dims, context.GetPlace());
+      set_zero(dev_ctx, &tmp, static_cast<T>(0));
+      auto tmp_t = framework::EigenTensor<T, D, Eigen::RowMajor,
+                                          Eigen::DenseIndex>::From(tmp);
+
+      grad_input_t.stridedSlice(starts_indices, ends_indices, steps_indices)
+          .device(place) = tmp_t;
+    }
+    if (grad_value) {
```
Comment (chenwhql) [inline paddle/fluid/operators/set_value_op.h:456]: do we need this if-cond here? it has been judged above

File: paddle/fluid/operators/set_value_op.h [hunk 4]
```diff
@@ -292,5 +312,255 @@ class SetValueKernel : public framework::OpKernel<T> {
...
+                                              Eigen::DenseIndex>::From(tmp);
+
+          tmp_t.device(place) =
+              in_t.stridedSlice(starts_indices, ends_indices, steps_indices);
+          grad_value_t.device(place) = tmp_t.reverse(reverse_axis);
+        } else {
+          grad_value_t.device(place) =
+              in_t.stridedSlice(starts_indices, ends_indices, steps_indices);
+        }
+      } else {
```
Comment (chenwhql) [inline paddle/fluid/operators/set_value_op.h:481]: what case grad_value->dims() != out_dims is ?

File: python/paddle/fluid/tests/unittests/test_set_value_op.py [hunk 5]
```diff
@@ -915,7 +917,315 @@ def test_dynamic(self):
...
+                'ValueTensor': value,
+                'StartsTensorList': [start, ],
+                'EndsTensorList': [end, ],
+                'StepsTensorList': [step, ]
+            }
+
+            helper = LayerHelper("set_value")
+            y = helper.create_variable_for_type_inference(dtype=x.dtype)
+
+            helper.append_op(
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_set_value_op.py:1114]: why need this form to test? form like `a[0] = value` can't work in static mode?

File: paddle/fluid/operators/set_value_op.h (hunk 4 above)
Comment (hbwx24) [reply paddle/fluid/operators/set_value_op.h:481]: eg: `t[0] = value`, `t[0].shape` is `[3,3]`, `value.shape` is `[1,]`,

File: paddle/fluid/operators/set_value_op.h (hunk 3 above)
Comment (hbwx24) [reply paddle/fluid/operators/set_value_op.h:456]: Done, thx.

File: paddle/fluid/operators/set_value_op.h (hunk 2 above)
Comment (hbwx24) [reply paddle/fluid/operators/set_value_op.h]: Done：`if ((grad_value) && (!grad_value->IsInitialized()))` -> `if(grad_value)`

File: paddle/fluid/operators/set_value_op.h (hunk 1 above)
Comment (hbwx24) [reply paddle/fluid/operators/set_value_op.h]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_set_value_op.py (hunk 5 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_set_value_op.py:1114]: set_value_op in __get/setitem__ is an inplace operation. When `input.stop_gradient = True` and `value.stop_gradient = False`,  set_value_grad_op will not be run during backward.

**Review by chenwhql — APPROVED** [review]
LGTM

**Pull request merged**